File: experiments/problem1_nonperiodicity/analyze_period_two_concrete_shadow_beliefs.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iostream>
#include <limits>
#include <map>
#include <set>
#include <stdexcept>
#include <string>
#include <tuple>
#include <unordered_map>
#include <utility>
#include <vector>

using u64 = std::uint64_t;
constexpr int MAXK = 25;
constexpr int SCHEDULE_CAP = 64;

u64 forward_generator(char generator, u64 state) {
    const u64 stepped = state ^ ((state << 1) | (state << 2));
    if (generator == 't') return stepped;
    if (generator == 'u') return stepped ^ 1;
    if (generator == 'p') return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    throw std::runtime_error("unknown generator");
}

std::vector<u64> next_level(const std::vector<u64>& current) {
    std::vector<u64> output;
    output.reserve(current.size() * 3);
    for (u64 state : current) {
        output.push_back(forward_generator('t', state));
        output.push_back(forward_generator('u', state));
        output.push_back(forward_generator('p', state));
    }
    std::sort(output.begin(), output.end());
    output.erase(std::unique(output.begin(), output.end()), output.end());
    return output;
}

std::vector<std::vector<u64>> build_levels(char phase, int maximum_complexity) {
    std::vector<std::vector<u64>> levels(maximum_complexity + 1);
    levels[1] = {phase == 'p' ? 3ULL : 1ULL};
    for (int k = 2; k <= maximum_complexity; ++k) {
        levels[k] = next_level(levels[k - 1]);
    }
    return levels;
}

bool contains(const std::vector<u64>& values, u64 target) {
    return std::binary_search(values.begin(), values.end(), target);
}

int bit_length(u64 value) {
    return value ? 64 - __builtin_clzll(value) : 0;
}

int expected_bits(char phase, int complexity) {
    return phase == 'p' ? 2 * complexity : 2 * complexity - 1;
}

std::string forced_zero_schedule(u64 state) {
    std::string word;
    for (int step = 0; step < SCHEDULE_CAP; ++step) {
        const u64 residue = state & 15;
        char branch;
        if (residue == 7) branch = 'u';
        else if (residue == 11) branch = 't';
        else return word;
        state = forward_generator(
            branch, forward_generator('p', (state - 3) >> 2)
        );
        word.push_back(branch);
    }
    throw std::runtime_error("forced schedule reached safety cap");
}

bool admissible(const std::string& word) {
    static const std::array<std::string, 3> forbidden{"uu", "ttttt", "ututtu"};
    for (const auto& factor : forbidden) {
        if (word.find(factor) != std::string::npos) return false;
    }
    return true;
}

std::string return_extension(const std::array<int, 3>& gaps, bool final_u) {
    std::string word = "u";
    for (int index = 0; index < 3; ++index) {
        word.append(gaps[index] - 1, 't');
        if (index < 2 || final_u) word.push_back('u');
    }
    return word;
}

struct Pattern {
    std::array<int, 3> gaps;
    std::string target;
    std::string complete;
};

std::vector<Pattern> three_return_patterns() {
    std::vector<Pattern> result;
    for (int a = 2; a <= 5; ++a) {
        for (int b = 2; b <= 5; ++b) {
            for (int c = 2; c <= 5; ++c) {
                std::array<int, 3> gaps{a, b, c};
                const std::string target = return_extension(gaps, false);
                const std::string complete = return_extension(gaps, true);
                if (admissible(complete)) result.push_back({gaps, target, complete});
            }
        }
    }
    return result;
}

std::vector<u64> direct_belief(
    const std::vector<std::vector<u64>>& levels,
    int complexity,
    u64 current,
    int depth
) {
    const u64 mask = (u64{1} << (2 * depth)) - 1;
    const u64 residue = current & mask;
    std::vector<u64> result;
    for (u64 state : levels[complexity - 1]) {
        if ((state & mask) == residue) result.push_back(state);
    }
    return result;
}

std::vector<u64> recursive_belief(
    const std::vector<std::vector<u64>>& levels,
    int complexity,
    u64 current,
    int depth
) {
    const int digit = current & 3;
    if (depth == 1) {
        std::vector<u64> result;
        for (u64 state : levels[complexity - 1]) {
            if ((state & 3) == u64(digit)) result.push_back(state);
        }
        return result;
    }
    auto parent = recursive_belief(levels, complexity - 1, current >> 2, depth - 1);
    std::vector<u64> result;
    result.reserve(parent.size());
    for (u64 state : parent) {
        const u64 child = 4 * state + unsigned(digit);
        if (contains(levels[complexity - 1], child)) result.push_back(child);
    }
    return result;
}

void verify_recursive_theorem() {
    for (char phase : {'p', 'u'}) {
        const auto levels = build_levels(phase, 9);
        for (int complexity = 2; complexity <= 9; ++complexity) {
            for (u64 current : levels[complexity]) {
                for (int depth = 1; depth < complexity; ++depth) {
                    if (direct_belief(levels, complexity, current, depth)
                        != recursive_belief(levels, complexity, current, depth)) {
                        throw std::runtime_error("direct/recursive belief mismatch");
                    }
                }
            }
        }
    }
}

struct Occurrence {
    u64 current;
    int cut;
    int depth;
    int weight;
};

struct Stats {
    u64 outputs = 0;
    u64 eligible_outputs = 0;
    u64 occurrences = 0;
    u64 positive_cut_occurrences = 0;
    u64 shadowed_occurrences = 0;
    u64 empty_beliefs = 0;
    u64 total_shadow_realizations = 0;
    u64 levels_with_occurrences = 0;
    u64 distinct_occurrence_cylinders = 0;
    int maximum_cut = 0;
    int maximum_depth = 0;
    u64 minimum_final_belief = std::numeric_limits<u64>::max();
    u64 maximum_final_belief = 0;
    int minimum_example_complexity = 0;
    int minimum_example_cut = 0;
    int minimum_example_depth = 0;
    u64 minimum_example_current = 0;
    u64 minimum_example_residue = 0;
    std::vector<u64> minimum_example_shadows;
};

Stats campaign(char phase, const std::vector<Pattern>& patterns) {
    const auto levels = build_levels(phase, MAXK);
    Stats stats;
    std::set<std::tuple<int, int, u64>> distinct;
    for (int k = 1; k <= MAXK; ++k) stats.outputs += levels[k].size();

    for (int complexity = 2; complexity <= MAXK; ++complexity) {
        std::vector<Occurrence> occurrences;
        std::set<int> needed_depths;
        for (u64 current : levels[complexity]) {
            if (bit_length(current) != expected_bits(phase, complexity)) {
                throw std::runtime_error("frontier bit-length law failed");
            }
            if ((current & 3) != 3) continue;
            ++stats.eligible_outputs;
            const std::string schedule = forced_zero_schedule(current);
            for (int cut = 0; cut <= static_cast<int>(schedule.size()); ++cut) {
                const std::string base = schedule.substr(0, cut);
                int weight = 0;
                for (const auto& pattern : patterns) {
                    if (schedule.compare(cut, pattern.target.size(), pattern.target) == 0
                        && admissible(base + pattern.complete)) {
                        ++weight;
                    }
                }
                if (!weight) continue;
                const int depth = cut + 1;
                if (depth >= complexity) {
                    throw std::runtime_error("occurrence consumed entire frontier");
                }
                occurrences.push_back({current, cut, depth, weight});
                needed_depths.insert(depth);
            }
        }
        if (occurrences.empty()) continue;
        ++stats.levels_with_occurrences;

        std::map<int, std::unordered_map<u64, u64>> counts;
        for (int depth : needed_depths) {
            const u64 mask = (u64{1} << (2 * depth)) - 1;
            auto& index = counts[depth];
            index.reserve(levels[complexity - 1].size() * 2);
            for (u64 state : levels[complexity - 1]) ++index[state & mask];
        }

        for (const auto& occurrence : occurrences) {
            const u64 mask = (u64{1} << (2 * occurrence.depth)) - 1;
            const u64 residue = occurrence.current & mask;
            const u64 belief_size = counts[occurrence.depth][residue];
            const u64 weight = occurrence.weight;
            stats.occurrences += weight;
            if (occurrence.cut > 0) stats.positive_cut_occurrences += weight;
            stats.maximum_cut = std::max(stats.maximum_cut, occurrence.cut);
            stats.maximum_depth = std::max(stats.maximum_depth, occurrence.depth);
            if (belief_size) stats.shadowed_occurrences += weight;
            else stats.empty_beliefs += weight;
            stats.total_shadow_realizations += weight * belief_size;
            if (belief_size < stats.minimum_final_belief) {
                stats.minimum_final_belief = belief_size;
                stats.minimum_example_complexity = complexity;
                stats.minimum_example_cut = occurrence.cut;
                stats.minimum_example_depth = occurrence.depth;
                stats.minimum_example_current = occurrence.current;
                stats.minimum_example_residue = residue;
                stats.minimum_example_shadows.clear();
                for (u64 state : levels[complexity - 1]) {
                    if ((state & mask) == residue) {
                        stats.minimum_example_shadows.push_back(state);
                    }
                }
                if (stats.minimum_example_shadows.size() != belief_size) {
                    throw std::runtime_error("minimum belief recount mismatch");
                }
            }
            stats.maximum_final_belief = std::max(stats.maximum_final_belief, belief_size);
            distinct.insert({complexity, occurrence.depth, residue});
        }
    }
    // Complexity one is eligible only for phase p and cannot contain a three-return occurrence.
    for (u64 state : levels[1]) if ((state & 3) == 3) ++stats.eligible_outputs;
    stats.distinct_occurrence_cylinders = distinct.size();
    return stats;
}

void print_stats(const std::string& prefix, const Stats& stats) {
    std::cout << prefix << "_outputs=" << stats.outputs << '\n';
    std::cout << prefix << "_eligible_outputs=" << stats.eligible_outputs << '\n';
    std::cout << prefix << "_occurrences=" << stats.occurrences << '\n';
    std::cout << prefix << "_positive_cut_occurrences=" << stats.positive_cut_occurrences << '\n';
    std::cout << prefix << "_shadowed_occurrences=" << stats.shadowed_occurrences << '\n';
    std::cout << prefix << "_empty_beliefs=" << stats.empty_beliefs << '\n';
    std::cout << prefix << "_total_shadow_realizations=" << stats.total_shadow_realizations << '\n';
    std::cout << prefix << "_levels_with_occurrences=" << stats.levels_with_occurrences << '\n';
    std::cout << prefix << "_distinct_occurrence_cylinders=" << stats.distinct_occurrence_cylinders << '\n';
    std::cout << prefix << "_maximum_cut=" << stats.maximum_cut << '\n';
    std::cout << prefix << "_maximum_depth=" << stats.maximum_depth << '\n';
    std::cout << prefix << "_minimum_final_belief=" << stats.minimum_final_belief << '\n';
    std::cout << prefix << "_maximum_final_belief=" << stats.maximum_final_belief << '\n';
    std::cout << prefix << "_minimum_example_complexity=" << stats.minimum_example_complexity << '\n';
    std::cout << prefix << "_minimum_example_cut=" << stats.minimum_example_cut << '\n';
    std::cout << prefix << "_minimum_example_depth=" << stats.minimum_example_depth << '\n';
    std::cout << prefix << "_minimum_example_current=0x" << std::hex << stats.minimum_example_current << std::dec << '\n';
    std::cout << prefix << "_minimum_example_residue=0x" << std::hex << stats.minimum_example_residue << std::dec << '\n';
    std::cout << prefix << "_minimum_example_shadows=";
    const std::size_t shown = std::min<std::size_t>(8, stats.minimum_example_shadows.size());
    for (std::size_t index = 0; index < shown; ++index) {
        if (index) std::cout << ',';
        std::cout << "0x" << std::hex << stats.minimum_example_shadows[index] << std::dec;
    }
    if (shown < stats.minimum_example_shadows.size()) std::cout << ",...";
    std::cout << '\n';
}

int main() {
    verify_recursive_theorem();
    const auto patterns = three_return_patterns();
    if (patterns.size() != 56) throw std::runtime_error("pattern count changed");
    const Stats phase_p = campaign('p', patterns);
    const Stats phase_u = campaign('u', patterns);

    std::cout << "maximum_complexity=" << MAXK << '\n';
    std::cout << "admissible_three_return_patterns=" << patterns.size() << '\n';
    std::cout << "recursive_theorem_small_exhaustive=1\n";
    print_stats("phase_p", phase_p);
    print_stats("phase_u", phase_u);
    std::cout << "total_outputs=" << phase_p.outputs + phase_u.outputs << '\n';
    std::cout << "total_eligible_outputs=" << phase_p.eligible_outputs + phase_u.eligible_outputs << '\n';
    std::cout << "total_occurrences=" << phase_p.occurrences + phase_u.occurrences << '\n';
    std::cout << "total_positive_cut_occurrences="
              << phase_p.positive_cut_occurrences + phase_u.positive_cut_occurrences << '\n';
    std::cout << "total_shadowed_occurrences="
              << phase_p.shadowed_occurrences + phase_u.shadowed_occurrences << '\n';
    std::cout << "total_empty_beliefs=" << phase_p.empty_beliefs + phase_u.empty_beliefs << '\n';
    std::cout << "total_shadow_realizations="
              << phase_p.total_shadow_realizations + phase_u.total_shadow_realizations << '\n';
    std::cout << "minimum_final_belief="
              << std::min(phase_p.minimum_final_belief, phase_u.minimum_final_belief) << '\n';
    std::cout << "maximum_final_belief="
              << std::max(phase_p.maximum_final_belief, phase_u.maximum_final_belief) << '\n';

    if (phase_p.outputs != 9118715 || phase_u.outputs != 7745997
        || phase_p.occurrences + phase_u.occurrences != 3395
        || phase_p.empty_beliefs + phase_u.empty_beliefs != 0) {
        throw std::runtime_error("complexity-25 totals changed");
    }
}

File: experiments/problem1_nonperiodicity/analyze_period_two_concrete_shadow_beliefs.py
#!/usr/bin/env python3
"""Exact residue-aware concrete shadow beliefs for period-two frontiers.

For phase a, current state x in O_(a,k), and cylinder depth L, define

    B_(a,k,L)(x) = {y in O_(a,k-1): y == x (mod 4^L)}.

This is the complete set of adjacent-complexity frontier realizations in the
same survivor cylinder.  Unlike a signature powerset, every member is one
actual concrete state, so digit updates cannot splice edges from unrelated
realizations.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from collections import Counter
from itertools import product
from typing import Any

PHASES = ("p", "u")
GAPS = (2, 3, 4, 5)
FORBIDDEN = ("uu", "ttttt", "ututtu")
DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 20
DEFAULT_SCHEDULE_CAP = 64


class ConcreteShadowLimitError(RuntimeError):
    """Raised before a controlled campaign exceeds its configured cap."""


def forward_generator(name: str, state: int) -> int:
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    return tuple(sorted({forward_generator(g, state) for g in "tup"}))


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def expected_bits(phase: str, complexity: int) -> int:
    return 2 * complexity if phase == "p" else 2 * complexity - 1


def build_levels(phase: str, maximum_complexity: int) -> list[set[int]]:
    levels: list[set[int]] = [set(), {phase_start(phase)}]
    for _ in range(2, maximum_complexity + 1):
        levels.append(
            {child for state in levels[-1] for child in frontier_children(state)}
        )
    return levels


def forced_zero_schedule(state: int, cap: int = DEFAULT_SCHEDULE_CAP) -> str:
    word: list[str] = []
    for _ in range(cap):
        residue = state & 15
        if residue == 7:
            branch = "u"
        elif residue == 11:
            branch = "t"
        else:
            return "".join(word)
        state = forward_generator(
            branch, forward_generator("p", (state - 3) >> 2)
        )
        word.append(branch)
    raise ConcreteShadowLimitError("forced schedule reached safety cap")


def admissible(word: str) -> bool:
    return not any(factor in word for factor in FORBIDDEN)


def return_extension(gaps: tuple[int, ...], include_final_u: bool) -> str:
    word = "u"
    for index, gap in enumerate(gaps):
        word += "t" * (gap - 1)
        if index < len(gaps) - 1 or include_final_u:
            word += "u"
    return word


def three_return_patterns() -> tuple[tuple[tuple[int, ...], str, str], ...]:
    rows = []
    for gaps in product(GAPS, repeat=3):
        target = return_extension(gaps, False)
        complete = return_extension(gaps, True)
        if admissible(complete):
            rows.append((gaps, target, complete))
    return tuple(rows)


def concrete_shadow_belief_direct(
    levels: list[set[int]], complexity: int, current: int, depth: int
) -> tuple[int, ...]:
    """Return every level-(k-1) state in the current depth-L cylinder."""
    if not 2 <= complexity < len(levels):
        raise ValueError("complexity outside built levels")
    if not 1 <= depth < complexity:
        raise ValueError("depth must satisfy 1 <= depth < complexity")
    if current not in levels[complexity]:
        raise ValueError("current state is not in the requested frontier")
    modulus = 4**depth
    residue = current % modulus
    return tuple(sorted(state for state in levels[complexity - 1] if state % modulus == residue))


def lift_belief_one_digit(
    levels: list[set[int]], target_level: int, belief: tuple[int, ...], digit: int
) -> tuple[int, ...]:
    """Append one common low base-four digit to one concrete belief."""
    if digit not in range(4):
        raise ValueError("digit must be in {0,1,2,3}")
    target = levels[target_level]
    return tuple(4 * state + digit for state in belief if 4 * state + digit in target)


def concrete_shadow_belief_recursive(
    levels: list[set[int]], complexity: int, current: int, depth: int
) -> tuple[int, ...]:
    """Compute the exact belief by recursive common-digit lifting.

    The base case selects the complete previous frontier with the required low
    digit.  For greater depth, strip the common digit from the current state,
    recurse one level lower, then append that exact digit to every surviving
    concrete shadow realization.
    """
    if not 2 <= complexity < len(levels):
        raise ValueError("complexity outside built levels")
    if not 1 <= depth < complexity:
        raise ValueError("depth must satisfy 1 <= depth < complexity")
    if current not in levels[complexity]:
        raise ValueError("current state is not in the requested frontier")
    digit = current & 3
    if depth == 1:
        return tuple(sorted(state for state in levels[complexity - 1] if state & 3 == digit))
    parent = current >> 2
    belief = concrete_shadow_belief_recursive(
        levels, complexity - 1, parent, depth - 1
    )
    return lift_belief_one_digit(levels, complexity - 1, belief, digit)


def concrete_shadow_trace(
    levels: list[set[int]], complexity: int, current: int, depth: int
) -> tuple[int, ...]:
    """Return belief cardinalities from the first common digit to full depth."""
    if depth == 1:
        return (len(concrete_shadow_belief_recursive(levels, complexity, current, 1)),)
    parent_trace = concrete_shadow_trace(levels, complexity - 1, current >> 2, depth - 1)
    full = concrete_shadow_belief_recursive(levels, complexity, current, depth)
    return parent_trace + (len(full),)


def phase_campaign(
    phase: str, maximum_complexity: int, schedule_cap: int
) -> dict[str, Any]:
    levels = build_levels(phase, maximum_complexity)
    patterns = three_return_patterns()
    totals = Counter(
        outputs=sum(len(levels[k]) for k in range(1, maximum_complexity + 1))
    )
    totals["eligible_outputs"] = sum(
        1
        for k in range(1, maximum_complexity + 1)
        for state in levels[k]
        if state & 3 == 3
    )
    belief_sizes: Counter[int] = Counter()
    belief_sizes_by_depth: Counter[tuple[int, int]] = Counter()
    distinct_cylinders: set[tuple[int, int, int]] = set()
    examples: list[dict[str, Any]] = []
    smallest: tuple[int, dict[str, Any]] | None = None

    for complexity in range(2, maximum_complexity + 1):
        level_had_occurrence = False
        for current in sorted(levels[complexity]):
            if current.bit_length() != expected_bits(phase, complexity):
                raise AssertionError("frontier bit-length law failed")
            if current & 3 != 3:
                continue
            schedule = forced_zero_schedule(current, schedule_cap)
            for cut in range(len(schedule) + 1):
                base = schedule[:cut]
                matches = [
                    gaps
                    for gaps, target, complete in patterns
                    if schedule[cut:].startswith(target)
                    and admissible(base + complete)
                ]
                if not matches:
                    continue
                level_had_occurrence = True
                depth = cut + 1
                if depth >= complexity:
                    raise AssertionError("occurrence cylinder consumed entire frontier")
                direct = concrete_shadow_belief_direct(
                    levels, complexity, current, depth
                )
                recursive = concrete_shadow_belief_recursive(
                    levels, complexity, current, depth
                )
                if direct != recursive:
                    raise AssertionError("direct and recursive beliefs disagree")
                trace = concrete_shadow_trace(levels, complexity, current, depth)
                residue = current % (4**depth)
                distinct_cylinders.add((complexity, depth, residue))
                weight = len(matches)
                totals["occurrences"] += weight
                totals["maximum_cut"] = max(totals["maximum_cut"], cut)
                totals["maximum_depth"] = max(totals["maximum_depth"], depth)
                if cut:
                    totals["positive_cut_occurrences"] += weight
                if direct:
                    totals["shadowed_occurrences"] += weight
                else:
                    totals["empty_beliefs"] += weight
                totals["total_shadow_realizations"] += weight * len(direct)
                belief_sizes[len(direct)] += weight
                belief_sizes_by_depth[(depth, len(direct))] += weight
                row = {
                    "complexity": complexity,
                    "state_hex": hex(current),
                    "cut": cut,
                    "depth": depth,
                    "residue_hex": hex(residue),
                    "gaps": [list(gaps) for gaps in matches],
                    "belief_size": len(direct),
                    "belief_trace": list(trace),
                    "first_shadow_hex": hex(direct[0]) if direct else None,
                }
                if smallest is None or len(direct) < smallest[0]:
                    smallest = (len(direct), row)
                if cut > 0 and len(examples) < 8:
                    examples.append(row)
        if level_had_occurrence:
            totals["levels_with_occurrences"] += 1

    if totals["occurrences"]:
        totals["distinct_occurrence_cylinders"] = len(distinct_cylinders)
        totals["minimum_final_belief"] = min(belief_sizes)
        totals["maximum_final_belief"] = max(belief_sizes)

    return {
        "phase": phase,
        "totals": dict(totals),
        "belief_size_histogram": {
            str(size): count for size, count in sorted(belief_sizes.items())
        },
        "belief_sizes_by_depth": {
            f"depth-{depth}/size-{size}": count
            for (depth, size), count in sorted(belief_sizes_by_depth.items())
        },
        "smallest_belief_example": smallest[1] if smallest else None,
        "positive_cut_examples": examples,
    }


def run_campaign(
    maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY,
    schedule_cap: int = DEFAULT_SCHEDULE_CAP,
) -> dict[str, Any]:
    if not 2 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise ConcreteShadowLimitError("maximum complexity outside controlled range")
    phases = {
        phase: phase_campaign(phase, maximum_complexity, schedule_cap)
        for phase in PHASES
    }
    additive_fields = (
        "outputs",
        "eligible_outputs",
        "occurrences",
        "positive_cut_occurrences",
        "shadowed_occurrences",
        "empty_beliefs",
        "total_shadow_realizations",
        "levels_with_occurrences",
        "distinct_occurrence_cylinders",
    )
    combined: dict[str, int] = {
        field: sum(phases[phase]["totals"].get(field, 0) for phase in PHASES)
        for field in additive_fields
    }
    combined["maximum_cut"] = max(
        phases[phase]["totals"].get("maximum_cut", 0) for phase in PHASES
    )
    combined["maximum_depth"] = max(
        phases[phase]["totals"].get("maximum_depth", 0) for phase in PHASES
    )
    combined["minimum_final_belief"] = min(
        phases[phase]["totals"]["minimum_final_belief"] for phase in PHASES
    )
    combined["maximum_final_belief"] = max(
        phases[phase]["totals"]["maximum_final_belief"] for phase in PHASES
    )
    payload: dict[str, Any] = {
        "status": "exact-residue-aware-concrete-shadow-belief-recursion",
        "maximum_complexity": maximum_complexity,
        "schedule_cap": schedule_cap,
        "admissible_three_return_patterns": len(three_return_patterns()),
        "theorem": {
            "belief_definition": (
                "B_(a,k,L)(x) is the complete set of O_(a,k-1) states congruent "
                "to x modulo 4^L. It is nonempty exactly when the cylinder has "
                "an adjacent-complexity concrete shadow."
            ),
            "recursive_update": (
                "For x=4q+d and L>=2, B_(a,k,L)(x) is obtained from "
                "B_(a,k-1,L-1)(q) by retaining exactly the concrete lifts "
                "4p+d that belong to O_(a,k-1)."
            ),
            "realization_consistency": (
                "Every recursive belief member is one actual frontier state; "
                "the update never combines transitions from unrelated states."
            ),
        },
        "phases": phases,
        "combined": combined,
        "scientific_boundary": (
            "The concrete belief recursion is exact at every depth. Nonemptiness "
            "for three-return occurrence cylinders is verified only through the "
            "configured complexity; this does not prove the all-depth adjacent-"
            "shadow inclusion, phase-complexity divergence, exclusion of eventual "
            "center period two, or Rule 30 center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity", type=int, default=DEFAULT_MAXIMUM_COMPLEXITY
    )
    parser.add_argument("--schedule-cap", type=int, default=DEFAULT_SCHEDULE_CAP)
    args = parser.parse_args()
    print(
        json.dumps(
            run_campaign(args.maximum_complexity, args.schedule_cap),
            indent=2,
            sort_keys=True,
        )
    )


if __name__ == "__main__":
    main()
